app/models/article.rb

```diff
@@ -1,4 +1,34 @@
 class Article < ApplicationRecord
   validates :title, presence: true
   validates :body, presence: true
+  validates :slug, uniqueness: true
+
+  before_save :set_slug
+
+  def to_param
+    self.slug || self.id
+  end
+
+  def set_slug
+    base_slug = self.title.parameterize(separator: "_")
+    counter = 1
+    current_slug = base_slug
+    while self.class.find_by(slug: current_slug)
+      current_slug = "#{base_slug}_#{counter.to_s}"
+      counter += 1
+    end
+    self.slug = current_slug
+  end
+
+  class << self
+    # Where "arg" could be either a string slug or an integer ID
+    def find_by_slug_or_id(arg)
+      if arg.to_i > 0
+        @article = Article.find(arg)
+      else
+        @article = Article.where(slug: arg).first
+      end
+    end
+
+  end
```

Comment:
Hey Sherwin! Not sure you need to be worrying about the uniqueness of slugs. As long as you enforce that the :id in /articles/:id can represent, for example, either 1023 or 1023-example-article-title, which both evaluate to 1023 when to_i is called on them. 

This is how it's done on Medium, if you check out the id-slug combination in their url. It's considered the better practice, since if you scale to millions of articles, uniqueness should rely on the id in the url, whereas the slug should serve as nothing more than a decorator in the URL.

Let me know if that makes sense.